feat: watch all grammars being edited

aleclarson opened the pull request:
Other changes:
  - Watch both .cson and .json grammar files
  - Only watch `language-*/grammars/*` paths
  - Only reload grammar packages with changes
  - Properly reload grammar packages
  - Live reloading is enabled by default
  - Support for Atom < v1.11.0 is removed
  - Improved debug logs
  - Simplified implementation

Fixes #3

aleclarson committed:
feat: watch all grammars being edited (no configuration)

Other changes:
  - Watch both .cson and .json grammar files
  - Only watch "language-*/grammars/*" paths
  - Only reload grammar packages with changes
  - Reload for "settings/*" changes
  - Properly reload grammar packages
  - Live reloading is enabled by default
  - Support for Atom < v1.11.0 is removed
  - Improved debug logs
  - Simplified implementation

ldez commented on lib/main.coffee:
```diff
@@ -1,114 +1,80 @@
 path = require 'path'
-promiseHelper = require './promise-helper'
+
+grammarRE = /\/language-[^\/]+\/grammars\/[^\/]+\.(?:c|j)son/
```
I prefer keep the configuration: because I don't want to reload all languages grammars in my workspace, but just one.
But it's possible to find way to do your needs and my needs.

aleclarson replied:
If you save a grammar `.cson` or `.json`, all affected editors will be refreshed. Only the changed grammar is reloaded. So I don't understand your complaint here.

ldez replied:
I want to choose which grammar I reload.
I have some grammar who don't this, because the reload system is integrated to the grammar package.

aleclarson replied:
Okay, I will add a whitelist config, but if empty, all grammars are watched. Does that sound good?

aleclarson replied:
Perhaps a blacklist config is more appropriate.

ldez replied:
Yes a blacklist seems good for me.

ldez commented on lib/main.coffee:
```diff
@@ -1,114 +1,80 @@
-    deactivate: ->
-      @subscriptions?.dispose()
-      @liveReloadSubscriptions?.dispose()
+    return unless atom.inDevMode()
```
Could you explain why you removed  `and not atom.inSpecMode()`

aleclarson replied:
No I can't. Can you explain what that did? I don't know what "spec mode" is or why it should be avoided.

ldez replied:
Spec mode it's the mode where the tests are run.

Could you restore this condition.

aleclarson replied:
Okay, that's what I suspected. But why shouldn't the package be enabled during tests? Perhaps tests should be added that actually test the core functionality.

ldez replied:
If you want add tests on the core functionality, it would make me happy, but it's not very simple

ldez replied:
I don't want the package be enable during the tests of others packages.

ldez commented on lib/main.coffee:
```diff
@@ -1,114 +1,80 @@
+  reload: (event) ->
+    {debug} = this
+
+    packPath = path.resolve event.path, '../..'
```
could you explain `'../..'`

aleclarson replied:
Imagine `event.path` is something like `/path/to/language-foo/grammars/foo.cson`.

The above line chops off the `/grammars/foo.cson` from the string, which leaves us with the package directory! 😄

ldez reviewed (changes requested):
Could you run and fix the tests.

aleclarson commented:
Are these the only tests? https://github.com/ldez/atom-grammar-live-reload/blob/master/spec/promise-helper-spec.coffee

ldez commented on lib/main.coffee:
```diff
@@ -1,114 +1,80 @@
+  for name, pack of atom.packages.loadedPackages
+    return pack if pack.path is path
+
+readCsonFile = (filePath) ->
```
Restore this method in this original place `promise-helper.coffee`.

aleclarson replied:
If we add integration tests, there's no need for `readCsonFile` to have its own unit test, and thus no need for its own file.

aleclarson committed:
feat: add blacklist config

aleclarson commented:
We're good to go!

I'm glad you didn't merge yet, because I noticed I could simplify the package reloading even more. 😄

I also realized the grammar settings weren't being reloaded, which resulted in issues like `⌘+/` not using the right syntax and indentation not working as expected. That has been fixed!

The blacklist feature is included in this pull request. You can put space-separated or comma-separated names, and it allows very sloppy spacing too! 👍

aleclarson committed:
fix: avoid watching same file twice

Editors of the same file share the same TextBuffer instance, so we only need to listen for save events once per file.

ldez reviewed (changes requested):
Your PR doesn't work me:
The language package path for dev package is not the real package path but a symbolic link into `/home/<USER>/.atom/dev/packages/`

ldez commented on lib/main.coffee:
```diff
@@ -1,114 +1,90 @@
+      reload = @reload.bind this
+      @editorSub = atom.workspace.observeTextEditors (editor) =>
+        filePath = editor.getPath()
+        if grammarRE.test(filePath) and not @watching[filePath]
```
I think the black list must be apply here.

aleclarson replied:
Oh that's a good point!

aleclarson commented:
For the symlink issue, I think the culprit is [`getPackageByPath`](https://github.com/aleclarson/atom-grammar-live-reload/blob/1286f490fc19f6a9130b190ea322c6dbda58449d/lib/main.coffee#L6-L8). We should check if `pkgPath` is a symlink (and resolve it) before calling that function. Would you mind helping out with that? 😄

ldez commented on lib/main.coffee:
```diff
@@ -1,114 +1,90 @@
+
+getPackageByPath = (path) ->
+  for name, pack of atom.packages.loadedPackages
+    return pack if pack.path is path
```
```coffee
return pack if path.endsWith pack.name
```

aleclarson replied:
Ah, yes! Very nice.

aleclarson replied:
`path.endsWith('/' + pack.name)` 😉

aleclarson commented:
Symlink issue fixed? Please try it again! 😄 

I also changed `grammarRE` to be `(grammars|settings)` so changes to the grammar settings cause a reload!

The blacklist is now checked before listeners are attached to an editor. We do so by looking at the package directory name (instead of `Package#name` which I think comes from `package.json`). So users need to be careful to specify the directory name and not the "name" field in `package.json`.

ldez commented on lib/main.coffee:
```diff
@@ -1,114 +1,97 @@
 path = require 'path'
-promiseHelper = require './promise-helper'
+
+grammarRE = /\/language-[^\/]+\/(grammars|settings)\/[^\/]+\.(?:c|j)son/
```
This must be set in the package configuration.

Also I think it will not work on Windows due to the Windows path separator `\`.

aleclarson replied:
What exactly do you think should be configurable?

ldez commented on lib/main.coffee:
```diff
@@ -1,114 +1,97 @@
+
+getPackageByPath = (path) ->
+  for name, pack of atom.packages.loadedPackages
+    return pack if path.endsWith '/' + pack.name
```
```coffee
getPackageByPath = (srcPath) ->
  for name, pack of atom.packages.loadedPackages
    return pack if path.basename(srcPath) is pack.name
```

aleclarson commented:
- Added [`toUnix`](https://github.com/aleclarson/atom-grammar-live-reload/blob/25aed4f931bd4143af703efaa9c51abd787e546f/lib/main.coffee#L6) to ensure `/` path separator is used before calling `grammarRE.test`

- Checked if [`path.resolve filePath, '../..'`](https://github.com/aleclarson/atom-grammar-live-reload/blob/25aed4f931bd4143af703efaa9c51abd787e546f/lib/main.coffee#L43) works with Windows paths (tested with `path.win32.resolve`) and it looks good

- Switched to using [`path.basename`](https://github.com/aleclarson/atom-grammar-live-reload/blob/25aed4f931bd4143af703efaa9c51abd787e546f/lib/main.coffee#L62-L63) for getting the `Package` instance, since that's a common assumption we're making (that the package directory name is equal to "name" in package.json)

- Added `$` to [`grammarRE`](https://github.com/aleclarson/atom-grammar-live-reload/blob/25aed4f931bd4143af703efaa9c51abd787e546f/lib/main.coffee#L4)

You think `grammarRE` should be configurable? Why?

aleclarson commented:
I think this is ready to merge. WDYT? 😄

ldez reviewed (approved):
LGTM
